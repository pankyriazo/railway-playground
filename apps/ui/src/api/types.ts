import { Deployments, Projects } from "@monorepo/types";

export type GetProjectListParams = {
  token: string;
};
export type GetProjectListResponse = Projects;

export type GetDeploymentListParams = {
  token: string;
  deploymentCount?: number;
  projectId: string;
  serviceId: string;
};
export type GetDeploymentListResponse = Deployments;
