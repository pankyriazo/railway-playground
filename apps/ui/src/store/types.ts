import { Deployment, Edges, Projects } from "@monorepo/types";

export type State = {
  token: string;
  projects: Projects;
  deployments: Record<string, Edges<Deployment>>;
};

interface ActionBase {
  type: string;
  payload?: unknown;
}

interface ResetAction extends ActionBase {
  type: "RESET";
}
interface SetTokenAction extends ActionBase {
  type: "SET_TOKEN";
  payload: string;
}
interface SetProjectsAction extends ActionBase {
  type: "SET_PROJECTS";
  payload: Projects;
}
interface AddDeploymentAction extends ActionBase {
  type: "ADD_DEPLOYMENT";
  payload: Record<string, Edges<Deployment>>;
}
interface RemoveDeploymentAction extends ActionBase {
  type: "REMOVE_DEPLOYMENT";
  payload: string;
}

export type Action =
  | ResetAction
  | SetTokenAction
  | SetProjectsAction
  | AddDeploymentAction
  | RemoveDeploymentAction;

export type Listener = (state: State) => void;

export type Reducer = (state: State, action: Action) => State;
