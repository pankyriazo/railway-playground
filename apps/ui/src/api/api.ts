import { SERVER_URL } from "@/environment";

const baseUrl = SERVER_URL;

const api = {
  getProjectList: (token: string) => {
    return fetch(`${baseUrl}/projects?token=${token}`).then((res) =>
      res.json()
    );
  },
};

export default api;
