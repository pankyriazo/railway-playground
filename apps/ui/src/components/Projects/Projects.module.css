.title {
  color: var(--color-text--600);
  font-size: var(--font-size-lg);
  font-weight: var(--font-weight-medium);
  margin-bottom: var(--spacing-md);
}

.skeletonContainer {
  display: flex;
  flex-direction: column;
  row-gap: 1rem;
}

.skeleton {
  composes: skeleton from global;

  border-radius: var(--border-radius);
  height: 10rem;
}

.projectList {
  display: flex;
  flex-direction: column;
  row-gap: var(--spacing-md);
}

.errorTitle {
  color: var(--color-text);
  font-size: var(--font-size-md);
  font-weight: var(--font-weight-medium);
}

.errorMessage {
  color: var(--color-text--800);
  font-size: var(--font-size-sm);
  font-weight: var(--font-weight-medium);
  margin-top: var(--spacing-sm);
}
