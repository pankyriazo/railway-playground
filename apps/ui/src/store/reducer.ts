import { Action, State } from "./types";
import { initialState } from "./store";

const reducer = (state: State, action: Action): State => {
  switch (action.type) {
    case "RESET":
      return initialState;
    case "SET_TOKEN":
      return { ...state, token: action.payload };
    case "SET_PROJECTS":
      return { ...state, projects: action.payload };
    case "ADD_DEPLOYMENT":
      return {
        ...state,
        deployments: { ...state.deployments, ...action.payload },
      };
    case "REMOVE_DEPLOYMENT":
      const { [action.payload]: _, ...deployments } = state.deployments;
      return { ...state, deployments };
    default:
      return state;
  }
};

export default reducer;
