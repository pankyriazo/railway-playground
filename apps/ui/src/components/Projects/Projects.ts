import api from "@/api";
import { Projects as ProjectsType } from "@monorepo/types";
import { StatefulComponent } from "@/types";
import store from "@/store";
import Project from "@/components/Project";

import styles from "./Projects.module.css";

class Projects extends StatefulComponent<HTMLDivElement> {
  private _token: string = "";
  private _projects: NonNullable<
    ProjectsType["data"]
  >["me"]["projects"]["edges"] = [];
  private _errorMessage: string = "";

  constructor() {
    super();

    this._element = document.createElement("div");
    this._element.id = "projects";

    this._subscribe();

    this.render();
  }

  private _subscribe() {
    return store.subscribe((state) => {
      if (state.token !== this._token) {
        this._token = state.token;
        this._fetchProjects();
      }

      if (
        state.projects.errors &&
        state.projects.errors[0] &&
        state.projects.errors[0].message !== this._errorMessage
      ) {
        this._errorMessage = state.projects.errors[0].message;
        this.setState("error");
      }

      if (
        state.projects.data &&
        state.projects.data.me.projects.edges !== this._projects
      ) {
        this._projects = state.projects.data.me.projects.edges;
        this.setState("loaded");
        this.render();
      }
    });
  }

  private _fetchProjects() {
    this.setState("loading");

    api
      .getProjectList({
        token: store.state.token,
      })
      .then((response) => {
        store.dispatch({
          type: "SET_PROJECTS",
          payload: response,
        });
      })
      .catch((e) => {
        this._errorMessage = e.message;
        this.setState("error");
      });
  }

  override render() {
    if (!this._element) {
      return;
    }

    this._element.innerHTML = "";
    this._element.dataset.testid = "projects";

    const title = document.createElement("h2");
    title.dataset.testid = "projects__title";
    title.className = styles.title!;
    title.textContent = "Projects";

    switch (this._state) {
      case "loading":
        this._element.appendChild(title);

        const skeletonContainer = document.createElement("div");
        skeletonContainer.dataset.testid = "projects__skeleton";
        skeletonContainer.className = styles.skeletonContainer!;

        for (let i = 0; i < 3; i++) {
          const skeleton = document.createElement("div");
          skeleton.className = styles.skeleton!;
          skeletonContainer.appendChild(skeleton);
        }

        this._element.appendChild(skeletonContainer);
        break;
      case "loaded":
        this._element.appendChild(title);

        const projectList = document.createElement("ul");
        projectList.dataset.testid = "projects__list";
        projectList.className = styles.projectList!;

        for (const projectData of this._projects) {
          const project = new Project(projectData.node);
          projectList.appendChild(project.element);
        }

        this._element.appendChild(projectList);
        break;
      case "error":
        this._element.appendChild(title);

        const errorTitle = document.createElement("h3");
        errorTitle.dataset.testid = "projects__error-title";
        errorTitle.textContent = "Error fetching projects";
        errorTitle.className = styles.errorTitle!;
        this._element.appendChild(errorTitle);

        const errorMessage = document.createElement("p");
        errorMessage.dataset.testid = "projects__error-message";
        errorMessage.textContent = this._errorMessage;
        errorMessage.className = styles.errorMessage!;
        this._element.appendChild(errorMessage);
        break;
    }
  }
}

export default Projects;
