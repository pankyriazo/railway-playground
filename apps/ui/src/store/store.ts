import reducer from "./reducer";
import { Action, Listener, Reducer, State } from "./types";

export const initialState: State = {
  token: "",
  projects: {
    data: null,
    errors: [],
  },
  deployments: {},
};

class Store {
  private _state: State;
  private _reducer: Reducer;
  private _listeners: Listener[] = [];

  constructor() {
    this._state = initialState;
    this._reducer = reducer;
  }

  get state(): State {
    return this._state;
  }

  private _notify(): void {
    this._listeners.forEach((listener) => listener(this._state));
  }

  dispatch(action: Action): void {
    this._state = this._reducer(this._state, action);
    this._notify();
  }

  subscribe(listener: Listener): () => void {
    this._listeners.push(listener);

    listener(this._state);

    return () => {
      this._listeners = this._listeners.filter((l) => l !== listener);
    };
  }
}

const store = new Store();

export default store;
