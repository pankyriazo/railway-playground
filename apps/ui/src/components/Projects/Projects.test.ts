import { mockComponent, render } from "@/test-helpers";
import {
  mockGetProjectList,
  mockGetProjectListError,
} from "@/api/__mocks__/getProjectList";
import api from "@/api";
import store from "@/store";
import Projects from "./Projects";

jest.mock("@/components/Project", () => mockComponent("Project"));
jest.mock("@/api", () => ({
  getProjectList: () => Promise.resolve({} as any),
}));

describe("Projects", () => {
  let apiSpy: jest.SpyInstance;

  beforeEach(() => {
    apiSpy = jest.spyOn(api, "getProjectList");

    render(Projects);
  });

  afterEach(() => {
    apiSpy.mockRestore();
  });

  describe("when the token is empty", () => {
    it("should not fetch projects", () => {
      expect(apiSpy).not.toHaveBeenCalled();
    });

    it("should not render the title", () => {
      const title = document.querySelector("[data-testid=projects__title]");

      expect(title).toBeNull();
    });

    it("should not render any projects", () => {
      const projects = document.querySelector("[data-testid=projects]");

      expect(projects!.innerHTML).toBe("");
    });
  });

  describe("when the token is updated", () => {
    beforeEach(() => {
      store.dispatch({
        type: "SET_TOKEN",
        payload: "test-token",
      });
    });

    it("should fetch projects", () => {
      expect(apiSpy).toHaveBeenCalled();
    });

    it("should render the title", () => {
      const title = document.querySelector("[data-testid=projects__title]");

      expect(title!.textContent).toBe("Projects");
    });

    it("should render the loading state", async () => {
      const projects = document.querySelector("[data-testid=projects]");

      expect(projects!.innerHTML).not.toBe("");
      expect(projects).toMatchSnapshot();
    });
  });

  describe("when the projects are loaded", () => {
    beforeEach(() => {
      store.dispatch({
        type: "SET_PROJECTS",
        payload: mockGetProjectList,
      });
    });

    it("should render the title", () => {
      const title = document.querySelector("[data-testid=projects__title]");

      expect(title!.textContent).toBe("Projects");
    });

    it("should render the projects", () => {
      const projectsElement = document.querySelector("[data-testid=projects]");

      const projects = document.querySelectorAll("[data-component=Project]");

      expect(projects.length).toBe(1);
      expect(projectsElement).toMatchSnapshot();
    });
  });

  describe("when there is an error fetching the projects", () => {
    beforeEach(() => {
      store.dispatch({
        type: "SET_PROJECTS",
        payload: mockGetProjectListError,
      });
    });

    it("should render the error state", () => {
      const projects = document.querySelector("[data-testid=projects]");

      const errorTitle = projects!.querySelector(
        "[data-testid=projects__error-title]"
      );
      const errorText = projects!.querySelector(
        "[data-testid=projects__error-message]"
      );

      expect(errorTitle!.textContent).toBe("Error fetching projects");
      expect(errorText!.textContent).toBe("Unauthorized");
      expect(projects).toMatchSnapshot();
    });
  });
});
