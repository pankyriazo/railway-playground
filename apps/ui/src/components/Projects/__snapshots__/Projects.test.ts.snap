// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Projects when the projects are loaded should render the projects 1`] = `
<div
  data-testid="projects"
  id="projects"
>
  <h2
    class="title"
    data-testid="projects__title"
  >
    Projects
  </h2>
  <ul
    class="projectList"
    data-testid="projects__list"
  >
    <div
      data-component="Project"
    />
  </ul>
</div>
`;

exports[`Projects when the token is updated should render the loading state 1`] = `
<div
  data-testid="projects"
  id="projects"
>
  <h2
    class="title"
    data-testid="projects__title"
  >
    Projects
  </h2>
  <div
    class="skeletonContainer"
    data-testid="projects__skeleton"
  >
    <div
      class="skeleton"
    />
    <div
      class="skeleton"
    />
    <div
      class="skeleton"
    />
  </div>
</div>
`;

exports[`Projects when there is an error fetching the projects should render the error state 1`] = `
<div
  data-testid="projects"
  id="projects"
>
  <h2
    class="title"
    data-testid="projects__title"
  >
    Projects
  </h2>
  <h3
    class="errorTitle"
    data-testid="projects__error-title"
  >
    Error fetching projects
  </h3>
  <p
    class="errorMessage"
    data-testid="projects__error-message"
  >
    Unauthorized
  </p>
</div>
`;
